usr.bin/tput/tput.1
.\"	$NetBSD: tput.1,v 1.15.6.1 2012/05/23 10:08:27 yamt Exp $
.\"
.\" Copyright (c) 1989, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)tput.1	8.2 (Berkeley) 3/19/94
.\"
.Dd April 5, 2012
.Dt TPUT 1
.Os
.Sh NAME
.Nm tput ,
.Nm clear
.Nd terminal capability interface
.Sh SYNOPSIS
.Nm
.Op Fl T Ar term
.Ar attribute
.Op Ar attribute-args
.Ar ...
.Sh DESCRIPTION
.Nm
makes terminal-dependent information available to users or shell
applications.
The options are as follows:
.Bl -tag -width Ds
.It Fl T
The terminal name as specified in the
.Xr terminfo 5
database, for example,
.Dq vt100
or
.Dq xterm .
If not specified,
.Nm
retrieves the
.Dq Ev TERM
variable from the environment.
.El
.Pp
.Nm
outputs a string if the
.Ar attribute
is of type string; a number if it is of type integer.
Otherwise,
.Nm
exits 0 if the terminal has the capability and 1 if it does not,
without further action.
.Pp
If the
.Ar attribute
is of type string, and takes arguments (e.g. cursor movement,
the terminfo
.Dq cup
sequence) the arguments are taken from the command line immediately
following the attribute.
.Pp
The following special attributes are available:
.Bl -tag -width Ar
.It clear
Clear the screen (the
.Xr terminfo 5
.Dq cl
sequence).
.It init
Initialize the terminal (the
.Xr terminfo 5
.Dq is2
sequence).
.It longname
Print the descriptive name of the user's terminal type.
.It reset
Reset the terminal (the
.Xr terminfo 5
.Dq rs1 , rs2 , rs3
and
.Dq rf
sequence).
.El
.Sh EXIT STATUS
The exit status of
.Nm
is based on the last attribute specified.
If the attribute is of type string or of type integer,
.Nm
exits 0 if the attribute is defined for this terminal type and 1
if it is not.
If the attribute is of type boolean,
.Nm
exits 0 if the terminal has this attribute, and 1 if it does not.
.Nm
exits 2 if any error occurred.
.Sh EXAMPLES
.Bl -tag -width "tput cm 6 11 DC 6" -compact
.It Li "tput cl cm 5 10"
clear the screen and goto line 5 column 10
.Pp
.It Li "tput cm 6 11 DC 6"
goto line 6 column 11 and delete 6 characters
.El
.Sh SEE ALSO
.Xr termcap 3 ,
.Xr terminfo 3 ,
.Xr terminfo 5
.Sh HISTORY
The
.Nm
command appears in
.Bx 4.4 .
.Sh BUGS
.Nm
can't really distinguish between different types of attributes.
.Pp
Not all terminfo entries contain the reset sequence, so using the init
sequence may be more useful.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1703 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	urtwn(4): Add a driver for Realtek RTL8188CU/RTL8192CU USB IEEE
		802.11b/g/n wireless network device, ported from OpenBSD.
		[nonaka 20120325]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]


usr.bin/tput/tput.c
/*	$NetBSD: tput.c,v 1.21 2011/10/04 11:02:32 roy Exp $	*/

/*-
 * Copyright (c) 1980, 1988, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1988, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)tput.c	8.3 (Berkeley) 4/28/95";
#endif
__RCSID("$NetBSD: tput.c,v 1.21 2011/10/04 11:02:32 roy Exp $");
#endif /* not lint */

#include <termios.h>

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <term.h>
#include <termcap.h>
#include <unistd.h>

static int    outc(int);
static void   usage(void) __dead;
static char **process(const char *, const char *, char **);

int
main(int argc, char **argv)
{
	int ch, exitval, n;
	char *term;
	const char *p, *s;

	term = NULL;
	while ((ch = getopt(argc, argv, "T:")) != -1)
		switch(ch) {
		case 'T':
			term = optarg;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (!term && !(term = getenv("TERM")))
		errx(2, "No terminal type specified and no TERM "
		    "variable set in the environment.");
	setupterm(term, 0, NULL);
	for (exitval = 0; (p = *argv) != NULL; ++argv) {
		switch (*p) {
		case 'c':
			if (!strcmp(p, "clear"))
				p = "clear";
			break;
		case 'i':
			if (!strcmp(p, "init")) {
				s = tigetstr("is1");
				if (s != NULL)
					tputs(s, 0, outc);
				p = "is2";
			}
			break;
		case 'l':
			if (!strcmp(p, "longname")) {
				(void)printf("%s\n", longname());
				continue;
			}
			break;
		case 'r':
			if (!strcmp(p, "reset")) {
				s = tigetstr("rs1");
				if (s != NULL)
					tputs(s, 0, outc);
				p = "rs2";
			}
			break;
		}
		if (((s = tigetstr(p)) != NULL && s != (char *)-1) ||
		    ((s = tgetstr(p, NULL)) != NULL))
			argv = process(p, s, argv);
		else if ((((n = tigetnum(p)) != -1 && n != -2 )||
			   (n = tgetnum(p)) != -1))
			(void)printf("%d\n", n);
		else {
			exitval = tigetflag(p);
			if (exitval == -1)
				exitval = !tgetflag(p);
			else
				exitval = !exitval;
		}

		if (argv == NULL)
			break;
	}
	return argv ? exitval : 2;
}

static char **
process(const char *cap, const char *str, char **argv)
{
	static const char errfew[] =
	    "Not enough arguments (%d) for capability `%s'";
	static const char erresc[] =
	    "Unknown %% escape `%c' for capability `%s'";
	char c, l;
	const char *p;
	int arg_need, p1, p2, p3, p4, p5, p6, p7, p8, p9;

	/* Count how many values we need for this capability. */
	arg_need = 0;
	p = str;
	while ((c = *p++) != '\0') {
		if (c != '%')
			continue;
		c = *p++;
		if (c == '\0')
			break;
		if (c != 'p')
			continue;
		c = *p++;
		if (c < '1' || c > '9')
			errx(2, erresc, c, cap);
		l = c - '0';
		if (l > arg_need)
			arg_need = l;
	}
	
#define NEXT_ARG							      \
	{								      \
		if (*++argv == NULL || *argv[0] == '\0')		      \
			errx(2, errfew, 1, cap);			      \
	}

	if (arg_need > 0) {
		NEXT_ARG;
		p1 = atoi(*argv);
	} else
		p1 = 0;
	if (arg_need > 1) {
		NEXT_ARG;
		p2 = atoi(*argv);
	} else
		p2 = 0;
	if (arg_need > 2) {
		NEXT_ARG;
		p3 = atoi(*argv);
	} else
		p3 = 0;
	if (arg_need > 3) {
		NEXT_ARG;
		p4 = atoi(*argv);
	} else
		p4 = 0;
	if (arg_need > 4) {
		NEXT_ARG;
		p5 = atoi(*argv);
	} else
		p5 = 0;
	if (arg_need > 5) {
		NEXT_ARG;
		p6 = atoi(*argv);
	} else
		p6 = 0;
	if (arg_need > 6) {
		NEXT_ARG;
		p7 = atoi(*argv);
	} else
		p7 = 0;
	if (arg_need > 7) {
		NEXT_ARG;
		p8 = atoi(*argv);
	} else
		p8 = 0;
	if (arg_need > 8) {
		NEXT_ARG;
		p9 = atoi(*argv);
	} else
		p9 = 0;

	/* And print them. */
	(void)tputs(tparm(str, p1, p2, p3, p4, p5, p6, p7, p8, p9), 0, outc);
	return argv;
}

static int
outc(int c)
{
	return putchar(c);
}

static void
usage(void)
{
	(void)fprintf(stderr,
	    "Usage: %s [-T term] attribute [attribute-args] ...\n",
	    getprogname());
	exit(2);
}
